godot-macros/src/docs.rs

```diff
@@ -0,0 +1,179 @@
+/// for godot's consumption.
+fn docs(doc: impl IntoIterator<Item = Attribute>) -> String {
+    let md = doc
+        .into_iter()
+        .filter_map(|x| {
+            let attr = x.path.first()?;
+            // find #[doc]
+            (attr.to_string() == "doc").then(|| {
+                match x.value {
+                    // #[doc = "…"]
+                    AttributeValue::Equals(_, doc) => Some(
+                        doc.into_iter()
+                            .map(|x| match x {
+                                // venial??
+                                proc_macro2::TokenTree::Literal(x) => x
+                                    .to_string()
+                                    .trim_start_matches('"')
+                                    .trim_end_matches('"')
+                                    .to_string(),
+                                _ => unreachable!(),
+                            })
+                            .reduce(|a, b| format!("{a} {b}"))
+                            .unwrap(),
+                    ),
+                    AttributeValue::Group(_, _) => None,
+                    // valid forms of #[doc] are
+                    // #[doc(hidden|inline|...)] or
+                    // #[doc = string]
+                    _ => unreachable!(),
+                }
+            })?
+        })
+        .collect::<String>();
+    md2bb::convert(&md)
+}
+
+fn signal(signal: SignalDefinition) -> String {
+    let name = signal.signature.name;
+    let params = params(signal.signature.params.iter().filter_map(|(x, _)| match x {
+        FnParam::Receiver(_) => None,
+        FnParam::Typed(y) => Some((y.name.clone(), y.ty.clone())),
+    }));
+    let desc = docs(signal.external_attributes);
+    format!(
+        r#"
+<signal name="{name}">
+  {params}
+  <description>
+  {desc}
+  </description>
+</signal>
+"#
+    )
+}
+
+fn constant(cnst: Constant) -> String {
+    let name = cnst.name.to_string();
+    let value = cnst
+        .initializer
+        .map(|x| x.to_token_stream().to_string())
+        .unwrap_or("null".into());
+    let docs = docs(cnst.attributes);
+    format!(r#"<constant name="{name}" value="{value}">{docs}</constant>"#)
+}
+
+pub fn member(member: Field) -> String {
+    let name = member.name.to_string();
+    let ty = member.ty.to_token_stream().to_string();
+    let docs = docs(member.attributes);
+    let default = member.default.to_token_stream().to_string();
+    format!(r#"<member name="{name}" type="{ty}" default="{default}">{docs}</member>"#)
+}
+
+fn params(params: impl Iterator<Item = (Ident, TypeExpr)>) -> String {
+    params
+        .zip(0..)
+        .fold(String::default(), |mut acc, ((name, ty), index)| {
+            use std::fmt::Write;
+            _ = write!(
+                acc,
+                r#"<param index="{index}" name="{name}" type="{ty}" />"#,
+                ty = ty.to_token_stream()
+            );
+            acc
+        })
```

Comment:
i feel like this would be simpler with just a `.map(..).collect()`

Reply:
yes, thats what i had, but clippy yelled at me, so...

Reply:
i had that, then clippy said that it was "unperformant" and i should use "write!" with a fold...

Reply:
really huh, i dislike that but if it really is? which lint is that?

Reply:
https://rust-lang.github.io/rust-clippy/master/index.html#/format_collect